modules/core/src/main/java/org/apache/synapse/util/xpath/SynapseExpression.java

```diff
@@ -15,104 +15,64 @@
-import org.apache.synapse.util.synapse.expression.ast.ExpressionNode;
-import org.apache.synapse.util.synapse.expression.ast.ExpressionResult;
-import org.apache.synapse.util.synapse.expression.context.EvaluationContext;
-import org.apache.synapse.util.synapse.expression.exception.EvaluationException;
-import org.apache.synapse.util.synapse.expression.exception.SyntaxErrorListener;
-import org.apache.synapse.util.synapse.expression.visitor.ExpressionVisitor;
 import org.jaxen.JaxenException;
 
-import java.util.HashMap;
-import java.util.Map;
+import java.lang.reflect.InvocationTargetException;
 
 /**
- * Represents a Synapse Expression
+ * Create a Synapse Expression by loading the implementation class.
  * Syntax ${ + expression +  } ex: expression="${vars.test + 5}"
  */
 public class SynapseExpression extends SynapsePath {
-    private static final Log log = LogFactory.getLog(SynapseExpression.class);
-    private final ExpressionNode expressionNode;
-    private final Map<String, String> namespaceMap = new HashMap<>();
-    private final boolean isContentAware;
 
-    public SynapseExpression(String synapseExpression) throws JaxenException {
-        super(synapseExpression, org.apache.synapse.config.xml.SynapsePath.JSON_PATH, log);
+    private final SynapsePath expression;
 
-        CharStream input = CharStreams.fromString(expression);
-        ExpressionLexer lexer = new ExpressionLexer(input);
-        CommonTokenStream tokens = new CommonTokenStream(lexer);
-        ExpressionParser parser = new ExpressionParser(tokens);
-        parser.removeErrorListeners();
-        SyntaxErrorListener errorListener = new SyntaxErrorListener();
-        parser.addErrorListener(errorListener);
+    private static final Log log = LogFactory.getLog(SynapseExpression.class);
 
-        ParseTree tree = parser.expression();
-        ExpressionVisitor visitor = new ExpressionVisitor();
-        expressionNode = visitor.visit(tree);
-        if (errorListener.hasErrors()) {
-            StringBuilder errorMessage = new StringBuilder("Syntax error in expression: " + synapseExpression);
-            for (SyntaxError error : errorListener.getErrors()) {
-                errorMessage.append(" ").append(error.getMessage());
+    public SynapseExpression(String synapseExpression) throws JaxenException {
+        super(synapseExpression, org.apache.synapse.config.xml.SynapsePath.SYNAPSE_EXPRESSIONS_PATH, log);
+        String expressionClass = SynapsePropertiesLoader.loadSynapseProperties().
+                getProperty(SynapseConstants.SYNAPSE_EXPRESSION_IMPL);
+        if (!StringUtils.isEmpty(expressionClass)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Trying to load the SynapseExpression implementation");
             }
-            throw new JaxenException(errorMessage.toString());
+            try {
+                Class<?> clazz = Class.forName(expressionClass);
```

Comment:
Dynamically loading the class per each expression may have a performance impact. Shall we try to optimise this?

Reply:
+1 fixed